p2panda-rs/src/test_utils/db/domain.rs

```diff
@@ -0,0 +1,1246 @@
+
+    // Verify that the claimed seq num matches the expected seq num for this author and log.
+    let latest_entry = store.get_latest_entry(author, log_id).await?;
+    let latest_seq_num = latest_entry.as_ref().map(|entry| entry.seq_num());
+    is_next_seq_num(latest_seq_num, seq_num)?;
+
+    // The backlink for this entry is the latest entry from this public key's log.
+    let backlink = latest_entry;
+
+    // If a skiplink is claimed, get the expected skiplink from the database, errors
+    // if it can't be found.
+    let skiplink = match entry.skiplink() {
+        Some(_) => Some(get_expected_skiplink(store, author, log_id, seq_num).await?),
+        None => None,
+    };
+
+    let skiplink_params: Option<(Entry, Hash)> = skiplink.map(|entry| {
+        let hash = entry.hash();
+        (entry.into(), hash)
+    });
+
+    let backlink_params: Option<(Entry, Hash)> = backlink.map(|entry| {
+        let hash = entry.hash();
+        (entry.into(), hash)
+    });
+
+    // Perform validation of the entry and it's operation.
+    let operation = validate_operation_with_entry(
+        &entry,
+        encoded_entry,
+        skiplink_params.as_ref().map(|(entry, hash)| (entry, hash)),
+        backlink_params.as_ref().map(|(entry, hash)| (entry, hash)),
+        plain_operation,
+        encoded_operation,
+        schema,
+    )?;
+
+    ///////////////////////////////////
+    // ENSURE SINGLE NODE PER AUTHOR //
+    ///////////////////////////////////
+
+    // @TODO: Missing a step here where we check if the author has published to this node before, and also
+    // if we know of any other nodes they have published to. Not sure how to do this yet.
+
+    //////////////////////////
+    // DETERMINE DOCUMENT ID //
+    //////////////////////////
+
+    let document_id = match operation.action() {
+        OperationAction::Create => {
+            // Derive the document id for this new document.
+            encoded_entry.hash().into()
+        }
+        _ => {
+            // We can unwrap previous operations here as we know all UPDATE and DELETE operations contain them.
+            let previous_operations = operation.previous_operations().unwrap();
+
+            // Get the document_id for the document_view_id contained in previous operations.
+            // This performs several validation steps (check method doc string).
+            let document_id =
+                get_checked_document_id_for_view_id(store, &previous_operations).await?;
+
+            // Ensure the document isn't deleted.
+            ensure_document_not_deleted(store, &document_id)
+                .await
+                .map_err(|_| {
+                    "You are trying to update or delete a document which has been deleted"
+                })?;
+
+            document_id
+        }
+    };
+
+    // Verify the claimed log id against the expected one for this document id and author.
+    verify_log_id(store, author, log_id, &document_id).await?;
```

Comment:
Maybe we can call this `verify_document_log_id` to not confuse it with entry / log id validation?